src/engine/srv.c

```diff
@@ -1280,6 +1280,39 @@ dss_srv_init(void)
 	return rc;
 }
 
+static void
+set_draining(void *arg)
+{
+	dss_get_module_info()->dmi_srv_shutting_down = 1;
+}
+
+/*
+ * Set the dss_module_info.dmi_srv_shutting_down flag for all xstreams, so that
+ * after this function returns, any dss_srv_shutting_down call (on any xstream)
+ * returns true. See also server_fini.
+ */
+void
+dss_srv_set_shutting_down(void)
+{
+	int	n = dss_xstream_cnt();
+	int	i;
+	int	rc;
+
+	/* Could be parallel... */
+	for (i = 0; i < n; i++) {
+		struct dss_xstream     *dx = dss_get_xstream(i);
+		ABT_task		task;
+
+		rc = ABT_task_create(dx->dx_pools[DSS_POOL_GENERIC], set_draining, NULL /* arg */,
+				     &task);
+		D_ASSERTF(rc == ABT_SUCCESS, "create task: %d\n", rc);
```

Comment:
I don't know if we do this elsewhere for ABT calls but if there's a way this can fail then we shouldn't really be using ASSERT.

Reply:
I assumed that this call can fail due to ENOMEM. Please note that we are responding to an engine shutdown request (e.g., a SIGTERM). And, if we abort the shutdown, daos_server will send a SIGKILL to us eventually, I think. If we ignore the error and continue the shutdown, we will fail other assertions or probably have worse consequences. Hence there doesn't seem to be a practically better way to handle this error.

Reply:
Not a lot else you can do here then, you could argue that you should catch this, log a D_CRIT and attempt to carry on but that's not really any better really.